design/local-node-agent.md

```diff
@@ -0,0 +1,147 @@
+   * This operation is synchronous and `Mount()` will not return until the mount/format is completed.
+
+#### Usage of CRDs
+The control flow described above explained how during the `Attach()` phase, the Rook agent (running on the Controller manager node) makes use of a volume attach CRD in order to convey to the agent running on the node that will consume the storage that there is a need to attach a volume to that node.
+This usage of the Kubernetes API is a major reason why the Flexvolume driver, which does not run in a context that has any cluster credentials, is insufficient for this design, and thus why a Rook agent pod is desirable.
+Let's enumerate some of the benefits of using a volume attach CRD:
+
+1. The admin gets a native `kubectl` experience for viewing and getting information on volume attachments.
+1. The CRD helps provide fencing for the volume in a generalized way, it is not specific to the underlying storage Rook is creating and managing.
+The existence of the CRD provides a means of bookkeeping to signal that the volume is locked and in use.
+1. A CRD removes any need for an API surface area on the agent.
+When the Flexvolume driver and Rook agent on the Controller manager node need to notify the agent on another node to perform the attach, a CRD provides a native Kubernetes means of communication, instead of the agent having to expose and secure a REST endpoint.
+
+In large scale clusters, it's important that the agent on each node only receives CRD events for CRD instances that are relevant to the node the agent is running on.
+To achieve this, each agent will be watching for volume attach CRD events using a `LabelSelector` for a `node=<node name>` label on the CRD instance.
+This filtering is performed on the server side, so only a single notification will be sent out to the agent on the relevant node.
+
+NOTE: Attempting to watch a CRD using a `LabelSelector` does not currently work and is being tracked by https://github.com/kubernetes/kubernetes/issues/51046.
+
+#### Improvements Over Rook's Current Support for Persistent Volumes
+Rook currently has a hybrid approach of providing persistent storage in a Kubernetes cluster.  While the Rook operator implements a dynamic volume provisioner, the attach and mounting is offloaded to the existing [RBD plugin](https://github.com/kubernetes/kubernetes/tree/master/pkg/volume/rbd).
+This requires that the Ceph tools are installed alongside the Kubelet on every node in the cluster, which causes friction in the experience for users of Rook.
+Furthermore, if users want to consume a Rook volume outside of the default namespace, then they must manually copy secrets to their namespace of choice.
+
+Both of these issues are handled by this proposed design for the Rook agent, greatly streamlining the Rook storage experience in Kubernetes and removing the common causes for errors and failure.  This is a big win for our users.
```

Comment:
i would also add that we are not tied to a K8S release. If we were to stay with the in-tree RBD plugin, any improvement we made to it, it would be only available on the following release.

Reply:
Added a note about being out-of-tree and the release timeline freedom that enables.